pkg/notifications/notifications.go

```diff
@@ -0,0 +1,447 @@
+func (s *Source) buildCommonEvent(rlr *redisLookupResult) (*event.Event, error) {
+	var (
+		eventName string
+		eventUrl  *url.URL
+		eventTags map[string]string
+	)
+
+	if rlr.ServiceName != "" {
+		eventName = rlr.HostName + "!" + rlr.ServiceName
+
+		eventUrl = s.notificationsClient.JoinIcingaWeb2Path("/icingadb/service")
+		eventUrl.RawQuery = "name=" + utils.RawUrlEncode(rlr.ServiceName) + "&host.name=" + utils.RawUrlEncode(rlr.HostName)
+
+		eventTags = map[string]string{
+			"host":    rlr.HostName,
+			"service": rlr.ServiceName,
+		}
+	} else {
+		eventName = rlr.HostName
+
+		eventUrl = s.notificationsClient.JoinIcingaWeb2Path("/icingadb/host")
+		eventUrl.RawQuery = "name=" + utils.RawUrlEncode(rlr.HostName)
+
+		eventTags = map[string]string{
+			"host": rlr.HostName,
+		}
+	}
+
+	return &event.Event{
+		Name: eventName,
+		URL:  eventUrl.String(),
+		Tags: eventTags,
+	}, nil
+}
+
+// buildStateHistoryEvent builds a fully initialized event.Event from a state history entry.
+//
+// The resulted event will have all the necessary information for a state change event, and must
+// not be further modified by the caller.
+func (s *Source) buildStateHistoryEvent(ctx context.Context, h *v1history.StateHistory) (*event.Event, error) {
+	res, err := s.fetchHostServiceName(ctx, h.HostId, h.ServiceId)
+	if err != nil {
+		return nil, err
+	}
+
+	ev, err := s.buildCommonEvent(res)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot build event for %q,%q", res.HostName, res.ServiceName)
+	}
+
+	ev.Type = event.TypeState
+
+	if res.ServiceName != "" {
+		switch h.HardState {
+		case 0:
+			ev.Severity = event.SeverityOK
+		case 1:
+			ev.Severity = event.SeverityWarning
+		case 2:
+			ev.Severity = event.SeverityCrit
+		case 3:
+			ev.Severity = event.SeverityErr
+		default:
+			return nil, fmt.Errorf("unexpected service state %d", h.HardState)
+		}
+	} else {
+		switch h.HardState {
+		case 0:
+			ev.Severity = event.SeverityOK
+		case 1:
+			ev.Severity = event.SeverityCrit
+		default:
+			return nil, fmt.Errorf("unexpected host state %d", h.HardState)
+		}
+	}
+
+	if h.Output.Valid {
+		ev.Message = h.Output.String
+	}
+	if h.LongOutput.Valid {
+		ev.Message += "\n" + h.LongOutput.String
+	}
+
+	return ev, nil
+}
+
+// buildDowntimeHistoryEvent from a downtime history entry.
+func (s *Source) buildDowntimeHistoryEvent(ctx context.Context, h *v1history.DowntimeHistory) (*event.Event, error) {
+	res, err := s.fetchHostServiceName(ctx, h.HostId, h.ServiceId)
+	if err != nil {
+		return nil, err
+	}
+
+	ev, err := s.buildCommonEvent(res)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot build event for %q,%q", res.HostName, res.ServiceName)
+	}
+
+	if h.HasBeenCancelled.Valid && h.HasBeenCancelled.Bool {
+		ev.Type = event.TypeDowntimeRemoved
+		ev.Message = "Downtime was cancelled"
+
+		if h.CancelledBy.Valid {
+			ev.Username = h.CancelledBy.String
+		}
+	} else if h.EndTime.Time().Compare(time.Now()) <= 0 {
```

Comment:
Huh, is comparing with now really the best option here? Shouldn't there be a more clear indication if this is a start or end event in the history entry?